src/storage.rs

```diff
@@ -44,133 +64,121 @@ pub trait ListableStore {
 
+/// A trait for a Zarr store that can write new data.
 pub trait WriteableStore {
+    /// The store writer.
     type SetWriter: Write;
 
+    /// Write a value to the Zarr store with the given `key` using a function `value` that yields
+    /// the item.
+    ///
+    /// # Errors
+    ///
+    /// A call to `set` may return an I/O error indicating the operation could not be completed.
     fn set<F: FnOnce(Self::SetWriter) -> Result<(), Error>>(
         &self,
         key: &str,
         value: F,
     ) -> Result<(), Error>;
 
-    // TODO differs from spec in that it returns a bool indicating existence of the key at the end of the operation.
+    /// Erase an item from the Zarr store.
+    ///
+    /// # Errors
+    ///
+    /// A call to `erase` may return an I/O error indicating the operation could not be completed.
     fn erase(&self, key: &str) -> Result<bool, Error>;
 
-    // TODO
+    /// Erase all items from the array store with a `key` containing the given `key_prefix`.
+    ///
+    /// # Errors
+    ///
+    /// A call to `erase_prefix` may return an I/O error indicating the operation could not be
+    /// completed.
     fn erase_prefix(&self, key_prefix: &str) -> Result<bool, Error>;
 }
 
-/// TODO
+/// Returns the chunk key of an array chunk at the given grid position.
+///
+/// # Examples
 ///
 /// ```
 /// use zarr::prelude::*;
-/// use zarr::storage::get_chunk_key;
 /// use zarr::smallvec::smallvec;
-/// let meta = ArrayMetadata::new(
-///     smallvec![50, 40, 30],
-///     smallvec![11, 10, 10],
-///     i8::ZARR_TYPE,
-///     zarr::compression::CompressionType::default(),
-/// );
-/// assert_eq!(get_chunk_key("/foo/baz", &meta, &[0, 0, 0]), "/data/root/foo/baz/c0/0/0");
-/// assert_eq!(get_chunk_key("/foo/baz", &meta, &[1, 2, 3]), "/data/root/foo/baz/c1/2/3");
-///
-/// let meta = ArrayMetadata::new(
-///     smallvec![],
-///     smallvec![],
-///     i8::ZARR_TYPE,
-///     zarr::compression::CompressionType::default(),
-/// );
-/// assert_eq!(get_chunk_key("/foo/baz", &meta, &[]), "/data/root/foo/baz/c");
+/// use zarr::storage::get_chunk_key;
+/// // Metadata for an array of shape (100, 4, 4) and chunk sizes of (100, 2, 2)
+/// let meta = ArrayMetadataBuilder::default()
+///     .shape(smallvec![100, 4, 4])
+///     .chunks(smallvec![100, 2, 2])
+///     .dtype(i8::ZARR_TYPE)
+///     .compressor(zarr::compression::CompressionType::default())
+///     .build()
+///     .unwrap();
+/// assert_eq!(get_chunk_key("/foo/baz", &meta, &[0, 0, 0]), "foo/baz/0.0.0");
+/// assert_eq!(get_chunk_key("/foo/baz", &meta, &[0, 1, 0]), "foo/baz/0.1.0");
+/// assert_eq!(get_chunk_key("/foo/baz", &meta, &[0, 0, 1]), "foo/baz/0.0.1");
+/// assert_eq!(get_chunk_key("/foo/baz", &meta, &[0, 1, 1]), "foo/baz/0.1.1");
 /// ```
-pub fn get_chunk_key(base_path: &str, array_meta: &ArrayMetadata, grid_position: &[u64]) -> String {
+#[must_use]
+pub fn get_chunk_key(key: &str, array_meta: &ArrayMetadata, grid_position: &[u64]) -> String {
```

Comment:
explore if we can get rid of this and any dependent potential v3-only code that we have alt implementations in private repo